scopes/harmony/pubsub/pubsub.preview.runtime.ts

```diff
@@ -5,60 +5,72 @@
 import { PreviewRuntime } from '@teambit/preview';
 import { isBrowser } from '@teambit/ui.is-browser';
 
+import { EventEmitter2 } from 'eventemitter2';
 import { connectToParent } from 'penpal';
 
 import { BitBaseEvent } from './bit-base-event';
 import { PubsubAspect } from './pubsub.aspect';
+import { Callback } from './types';
 
 type ParentMethods = {
-  sub: (topicUUID: string, event: any) => Promise<void>;
-  pub: (topicUUID: string, event: any) => Promise<void>;
+  pub: (topic: string, event: BitBaseEvent<any>) => Promise<any>;
 };
 
 export class PubsubPreview {
   private _parentPubsub?: ParentMethods;
+  private events = new EventEmitter2();
+
+  public sub(topic: string, callback: Callback) {
+    const emitter = this.events;
+    emitter.on(topic, callback);
+
+    const unSub = () => {
+      emitter.off(topic, callback);
+    };
+    return unSub;
+  }
+
+  public pub(topic: string, event: BitBaseEvent<any>) {
+    this.events.emit(topic, event);
+    this._parentPubsub?.pub(topic, event).catch(() => {});
+  }
 
   private inIframe() {
     try {
       return isBrowser && window.self !== window.top;
     } catch (e) {
-      return true;
+      return false;
     }
   }
 
-  private connectToParentPubSub = (retries = 10) => {
-    if (retries <= 0) return undefined;
+  private connectToParentPubSub = (retries = 10): Promise<ParentMethods | undefined> => {
+    if (retries <= 0) throw new Error('Pubsub.preview: could not connect to parent window');
 
-    return connectToParent<ParentMethods>({ timeout: 300 })
-      .promise.then((parentPubsub) => {
-        return (this._parentPubsub = parentPubsub);
-      })
+    return connectToParent<ParentMethods>({
+      timeout: 300,
+      methods: {
+        pub: this.handleMessageFromParent,
+      },
+    })
+      .promise.then((parentPubsub) => (this._parentPubsub = parentPubsub))
       .catch((e: any) => {
-        if (e.code !== 'ConnectionTimeout') return undefined;
+        if (e.code !== 'ConnectionTimeout') throw e;
 
         return this.connectToParentPubSub(retries - 1);
       });
   };
 
-  public sub(topicUUID, callback) {
-    if (!this._parentPubsub) return undefined;
-
-    return this._parentPubsub?.sub(topicUUID, callback);
-  }
-
-  public pub(topicUUID, event: BitBaseEvent<any>) {
-    if (!this._parentPubsub) return undefined;
-
-    return this._parentPubsub.pub(topicUUID, event);
-  }
+  private handleMessageFromParent = (topic: string, message: BitBaseEvent<any>) => {
+    this.events.emit(topic, message);
+  };
 
   static runtime = PreviewRuntime;
 
-  static async provider() {
+  static async provider(): Promise<PubsubPreview> {
     const pubsubPreview = new PubsubPreview();
 
     if (pubsubPreview.inIframe()) {
-      return pubsubPreview.connectToParentPubSub();
+      pubsubPreview.connectToParentPubSub().catch(() => {});
```

Comment:
please console the error here, otherwise, it will be very hard to understand an issue when it happens